waffle_hub/hub/hub.py

```diff
@@ -968,146 +1100,150 @@ def train(
-            for key in cfg.advance_params.keys():
-                if key not in default_advance_param:
+                        raise FileNotFoundError(f"Advance parameter file is not exist.")
+                elif not isinstance(cfg.advance_params, dict):
                     raise ValueError(
-                        f"Advance parameter {key} is not supported.\n"
-                        + f"Supported parameters: {list(default_advance_param.keys())}"
+                        f"Advance parameter should be dictionary or file path. {cfg.advance_params}"
                     )
 
-        callback = TrainCallback(cfg.epochs + 1, self.get_metrics)
-        result = TrainResult()
-        result.callback = callback
+                default_advance_param = self.get_default_advance_train_params()
+                for key in cfg.advance_params.keys():
+                    if key not in default_advance_param:
+                        raise ValueError(
+                            f"Advance parameter {key} is not supported.\n"
+                            + f"Supported parameters: {list(default_advance_param.keys())}"
+                        )
 
-        # TODO: hold arguemnt will be deprecated
-        if hold:
-            inner(callback, result)
-        else:
-            thread = threading.Thread(target=inner, args=(callback, result), daemon=True)
-            callback.register_thread(thread)
-            callback.start()
+            result = TrainResult()
+            status_controller.set_total_step(cfg.epochs)
+
+            # train run
+            status_controller.set_running()
+
+            metric_logger.start()
+            self.before_train(cfg)
+            self.on_train_start(cfg)
+            self.save_train_config(cfg)
+            self.training(cfg, status_controller)
+            self.on_train_end(cfg)
+            status_controller.set_success()
+            self.evaluate(
+                dataset=dataset,
+                batch_size=cfg.batch_size,
+                image_size=cfg.image_size,
+                letter_box=cfg.letter_box,
+                device=cfg.device,
+                workers=cfg.workers,
+            )
+            self.after_train(cfg, result)
+            metric_logger.stop()
+        except FileExistsError as e:
+            status_controller.set_failed(e)
+            if hasattr(metric_logger, "thread"):
+                metric_logger.stop()
+            raise e
+        except KeyboardInterrupt as e:
+            status_controller.set_stopped(e)
+            raise e
+        except SystemExit as e:
+            status_controller.set_stopped(e)
+            raise e
+        except Exception as e:
+            status_controller.set_failed(e)
+            if self.artifact_dir.exists():
+                io.remove_directory(self.artifact_dir)
+            if hasattr(metric_logger, "thread"):
+                metric_logger.stop()
+            raise e
```

Comment:
We can group multiple Exceptions in one line according to our business logic.

https://docs.python.org/3/tutorial/errors.html#handling-exceptions

```python
except (RuntimeError, TypeError, NameError):
    pass
```

And I have a question.
Why do the each hub adapter have same exception logic?
Isn't it enough to handle exceptions here?

Reply:
Sorry, It is my mistake :smiling_face_with_tear: ;;